Feature/service manager for rpc api

Ocramius opened the pull request:
This PR introduces following new features:
1.  service names mapped as JS API via config (See #4)
2.  caching enabled by default (using the memory adapter/dummy)
3.  deprecation of previous "crawl" logic in favor of direct explicit mapping (functionality is kept for BC)
4.  simplification of the `ModuleApi` class, which now is a simple wrapper of available actions (may rename it shortly into `ApiContainer` or `ActionContainer`)
5.  removal of requirement to use the view model produced by `KJSencha\Frontend\Bootstrap` in favor of explicit view helpers
6.  documentation of the introduced features

Hope you like it :D

Ocramius committed:
Reducing coupling of the controller with the ServiceLocator

Ocramius committed:
Adding basic logic (namespacing yet to be done)

Ocramius committed:
Working revision of mapping of services->direct api. Cache broken

Ocramius committed:
Making API objects serializable

Ocramius committed:
Fixing service/object names according to new standards

Ocramius committed:
Adding test echo service

Ocramius committed:
Documenting and simplifying view helpers to make KJSencha\Frontend\Bootstrap->getViewModel() deprecated

Ocramius committed:
Minor cleanups, IDE hinted code inspections

Ocramius committed:
Minor cleanups, IDE hinted code inspections

Ocramius committed:
Minor cleanups, IDE hinted code inspections

Ocramius committed:
Minor cleanups, IDE hinted code inspections

Rovak commented on config/services.config.php:
```diff
@@ -52,16 +70,28 @@
         },
+
+        /**
+         * Direct manager, handles instantiation of requested services
```
Duplicate comment

Rovak commented on src/KJSencha/View/Helper/Variables.php:
```diff
@@ -0,0 +1,42 @@
+        $script = '';
+
+        foreach ($this->bootstrap->getVariables() as $key => $value) {
+            $script .= 'var ' . $key . '=' . json_encode($value) . ';' . PHP_EOL;
```
What should we do with invalid javascript variable names that contain dots or slashes, should they be sanitized by our ViewHelper or just throw an error on the Javascript side. 
I vote for a simple preg_replace which filters nonalpha characters which should be sufficient.

Ocramius replied:
I guess that should be handled by `KJSencha\Frontend\Bootstrap` at set-time and validated... That's the easiest way to avoid having users do unexpected things, and is also quite self-documenting :)

Rovak commented on docs/intro.md:
````diff
@@ -0,0 +1,119 @@
+```
+
+## Enabling
+To enable KJSencha, open your `config/application.config.php` file and add following items
````
add `the` following items

Rovak commented on docs/intro.md:
```diff
@@ -0,0 +1,119 @@
+
+## View
+
+In your view (presumably `module/Application/view/application/index/index.phtml`), add
```
add `the`

Rovak commented on src/KJSencha/Direct/Remoting/Api/Factory/ModuleFactory.php:
```diff
@@ -2,66 +2,173 @@
+        return $action;
+    }
+
+    protected function buildMethod(MethodScanner $classMethod)
```
Missing method documentation

Rovak commented on src/KJSencha/Direct/Remoting/Api/Object/Action.php:
```diff
@@ -2,12 +2,16 @@
 
 namespace KJSencha\Direct\Remoting\Api\Object;
 
+use Zend\Stdlib\ArrayUtils;
```
Added but seems to be unused

Rovak commented on src/KJSencha/Direct/Remoting/Api/Object/Method.php:
```diff
@@ -2,11 +2,16 @@
 
 namespace KJSencha\Direct\Remoting\Api\Object;
 
+use Zend\Stdlib\ArrayUtils;
```
Seems to be unused

Rovak commented on src/KJSencha/Frontend/Direct/RemotingProvider.php:
```diff
@@ -14,7 +14,6 @@ class RemotingProvider extends Ext\Base
     protected $attributes = array(
         'type'		=> '',
         'url'		=> '',
-        'namespace' => '',
```
namespace is a required value and should stay there as a hint, maybe add a getter/setter to make it more clear like setUrl

Ocramius replied:
I removed `namespace` on JS-side :)

Rovak commented:
The PR looks really good, lots of improvements made in the overall code.

I don't agree with the deprecation of the crawl logic, the functionality is provided by a lot of other Ext.Direct generators and personally i find it really useful to be able to use convention over configuration.

What is your intended use for the `docs` folder? I prefer to keep the documentation centralized in the wiki, i shall cleanup the current mess tomorrow so it can be added there.

Ocramius commented:
@Rovak I prefer keeping docs with the repo itself since it forces users to clone it together... I can move it if you prefer so :)

On the crawl logic, the problem is that the only way to operate with (let's say) DB data on invokables are either service initializers (making invokables implement `SomethingAwareInterface`, which is kinda crappy) or using static methods, which is a "game over" for me (once you got the static method in, good luck getting it out of there!). You can see some of those patterns applied in the old view helpers, like [`Zend_View_Helper_Action`](http://framework.zend.com/svn/framework/standard/trunk/library/Zend/View/Helper/Action.php), where you just asked for an "action" plugin and the framework did all the inflection/file matching/autoloading things for you.
Direct mapping via services allows instead to avoid that by following the patterns introduced in ZF2.

It is mainly related to educating the end user about good and bad approaches. The technique of crawling directories for invokables is a niche and you meet its limitations too easily :(

Rovak commented:
For now i prefer to keep things in the Wiki, i may setup a Jenkins job later which syncs the `docs` files to the Wiki because i have to agree that cloning the docs together is helpful.

About the crawling, in our internal version we use a AbstractDirectService which is a lightweight AbstractController with a pluginmanager that contains ExtJS specific helpers. I mainly use Direct for really small PHP methods which call service methods, so i may not face the same problems as you are having.

There is a lot of room for improvement and i will create a new PR in which we can discuss further changes in the crawling as soon as this one is merged.

Ocramius commented:
@Rovak fine by me :)
I'm not saying "let's remove it". If you know what you are doing, then it's already fine. As said, it's more a matter of educating the end user. Will move on addressing your comments tomorrow, and then adding a new commit to remove any unused files/methods (to be reviewed, since it may be BC on your side)

Ocramius committed:
Addressing issues found by @Rovak's CR

Ocramius committed:
Removing unused objects and method, completing missing API docs

Ocramius commented on config/services.config.php:
```diff
@@ -33,10 +36,21 @@
+            /* @var $annotationManager AnnotationManager */
+            $annotationManager = $sl->get('kjsencha.annotationmanager');
+            /* @var $directManager DirectManager */
+            $directManager = $sl->get('kjsencha.annotationmanager');
```
I messed this line up, fixing.

Ocramius committed:
Renaming ModuleApi -> Api and ModuleFactory -> ApiBuilder

Ocramius committed:
Fixing service id messed up in previous work

Ocramius committed:
Automated CS fixes
